framework/src/modules/interoperability/base_cross_chain_update_command.ts

```diff
@@ -0,0 +1,241 @@
+		const baseStateSnapshotID = context.stateStore.createSnapshot();
+
+		try {
+			for (const [module, method] of this.interoperableCCMethods.entries()) {
+				if (method.beforeCrossChainCommandExecute) {
+					logger.debug(
+						{
+							moduleName: module,
+							commandName: ccm.crossChainCommand,
+							ccmID: ccmID.toString('hex'),
+						},
+						'Execute beforeCrossChainCommandExecute',
+					);
+					await method.beforeCrossChainCommandExecute(context);
+				}
+			}
+		} catch (error) {
+			context.eventQueue.restoreSnapshot(baseEventSnapshotID);
+			context.stateStore.restoreSnapshot(baseStateSnapshotID);
+			logger.info(
+				{ err: error as Error, moduleName: ccm.module, commandName: ccm.crossChainCommand },
+				'Fail to execute beforeCrossChainCommandExecute.',
+			);
+			await internalMethod.terminateChainInternal(ccm.sendingChainID, context);
+			this.events.get(CcmProcessedEvent).log(context, ccm.sendingChainID, ccm.receivingChainID, {
+				ccmID,
+				code: CCM_PROCESSED_CODE_INVALID_CCM_BEFORE_CCC_EXECUTION_EXCEPTION,
+				result: CCM_PROCESSED_RESULT_DISCARDED,
+			});
+			return;
+		}
+
+		const execEventSnapshotID = context.eventQueue.createSnapshot();
+		const execStateSnapshotID = context.stateStore.createSnapshot();
+
+		try {
+			await command.execute(context);
+			this.events.get(CcmProcessedEvent).log(context, ccm.sendingChainID, ccm.receivingChainID, {
+				ccmID,
+				code: CCM_PROCESSED_CODE_SUCCESS,
+				result: CCM_PROCESSED_RESULT_APPLIED,
+			});
+		} catch (error) {
+			context.eventQueue.restoreSnapshot(execEventSnapshotID);
+			context.stateStore.restoreSnapshot(execStateSnapshotID);
+			await this.bounce(
+				context,
+				ccmID,
+				encodedCCM.length,
+				CCM_STATUS_CODE_FAILED_CCM,
+				CCM_PROCESSED_CODE_FAILED_CCM,
+			);
+		}
+
+		try {
+			for (const [module, method] of this.interoperableCCMethods.entries()) {
+				if (method.afterCrossChainCommandExecute) {
+					logger.debug(
+						{
+							moduleName: module,
+							commandName: ccm.crossChainCommand,
+							ccmID: ccmID.toString('hex'),
+						},
+						'Execute afterCrossChainCommandExecute',
+					);
+					await method.afterCrossChainCommandExecute(context);
+				}
+			}
+		} catch (error) {
+			context.eventQueue.restoreSnapshot(baseEventSnapshotID);
+			context.stateStore.restoreSnapshot(baseStateSnapshotID);
+			logger.info(
+				{ err: error as Error, moduleName: module, commandName: ccm.crossChainCommand },
+				'Fail to execute afterCrossChainCommandExecute',
+			);
+			await internalMethod.terminateChainInternal(ccm.sendingChainID, context);
+			this.events.get(CcmProcessedEvent).log(context, ccm.sendingChainID, ccm.receivingChainID, {
+				ccmID,
+				code: CCM_PROCESSED_CODE_INVALID_CCM_AFTER_CCC_EXECUTION_EXCEPTION,
+				result: CCM_PROCESSED_RESULT_DISCARDED,
+			});
+		}
+	}
+
+	protected async bounce(
+		context: CrossChainMessageContext,
+		ccmID: Buffer,
```

Comment:
maybe `ccmID`, `ccmSize`, `ccmStatus` can be pulled from `context.ccm` so we don't need them as parameters here

Reply:
those are added because when using bounce, it's already computed value, and re-computing has some  costs of encode / hash

Reply:
ccmStatus is the status we want to put when bouncing, and not the same as ccm.status